### PR TITLE
chore(types): improve typing

### PR DESCRIPTION
**Pull request by kumaraditya303**
Fixes https://github.com/microsoft/playwright-python/issues/788

### CODE REVIEW

File: playwright/_impl/_async_base.py [hunk 1]
```diff
@@ -68,17 +73,19 @@ def _wrap_handler(self, handler: Any) -> Callable[..., None]:
             return mapping.wrap_handler(handler)
         return handler
 
-    def on(self, event: str, f: Any) -> None:
+    def on(self, event: str, f: Callable[..., Union[Awaitable[None], None]]) -> None:
```
Comment (kumaraditya303) [inline playwright/_impl/_async_base.py:76]: This will accept both sync and async event listeners

File: playwright/_impl/_sync_base.py [hunk 2]
```diff
@@ -110,17 +115,17 @@ def _wrap_handler(self, handler: Any) -> Callable[..., None]:
             return mapping.wrap_handler(handler)
         return handler
 
-    def on(self, event: str, f: Any) -> None:
+    def on(self, event: str, f: Callable[..., None]) -> None:
```
Comment (kumaraditya303) [inline playwright/_impl/_sync_base.py:118]: Accepts a Callable

File: playwright/_impl/_connection.py [hunk 3]
```diff
@@ -319,11 +319,12 @@ def _replace_guids_with_channels(self, payload: Any) -> Any:
         return payload
 
 
-def from_channel(channel: Channel) -> Any:
+def from_channel(channel: Channel) -> ChannelOwner:
```
Comment (kumaraditya303) [inline playwright/_impl/_connection.py:322]: As all inherit from ChannelOwner we eliminate runtime casts

File: playwright/_impl/_async_base.py (hunk 1 above)
Comment (pavelfeldman) [reply playwright/_impl/_async_base.py:76]: Let's only do one thing at a time in each PR. If this PR is about `cast`, it should only touch `cast`.

Comment (kumaraditya303) [reply playwright/_impl/_async_base.py:76]: Okay

File: playwright/_impl/_artifact.py [hunk 4]
```diff
@@ -37,7 +37,7 @@ async def path_after_finished(self) -> Optional[pathlib.Path]:
         return pathlib.Path(await self._channel.send("pathAfterFinished"))
 
     async def save_as(self, path: Union[str, Path]) -> None:
-        stream = cast(Stream, from_channel(await self._channel.send("saveAsStream")))
+        stream: Stream = from_channel(await self._channel.send("saveAsStream"))
```
Comment (pavelfeldman) [inline playwright/_impl/_artifact.py:40]: I just tried following locally and mypy was happy...

```py
stream: Stream = self
```

File: playwright/_impl/_async_base.py [hunk 5]
```diff
@@ -15,19 +15,19 @@
...
+from typing import Any, Awaitable, Callable, Generic, Type, TypeVar, Union
 
 from playwright._impl._impl_to_api_mapping import ImplToApiMapping, ImplWrapper
 
 mapping = ImplToApiMapping()
 
 
 T = TypeVar("T")
-Self = TypeVar("Self", bound="AsyncBase")
+Self = TypeVar("Self", bound="AsyncContextManager")
```
Comment (pavelfeldman) [inline playwright/_impl/_async_base.py:26]: Let's only deal with cast here.

File: playwright/_impl/_async_base.py [hunk 6]
```diff
@@ -15,19 +15,19 @@
...
 
 
 class AsyncEventInfo(Generic[T]):
-    def __init__(self, future: asyncio.Future) -> None:
+    def __init__(self, future: "asyncio.Future[T]") -> None:
```
Comment (pavelfeldman) [inline playwright/_impl/_async_base.py:30]: Let's only deal with cast here.

File: playwright/_impl/_async_base.py [hunk 7]
```diff
@@ -39,13 +39,18 @@ def is_done(self) -> bool:
 
 
 class AsyncEventContextManager(Generic[T]):
-    def __init__(self, future: asyncio.Future) -> None:
-        self._event: AsyncEventInfo = AsyncEventInfo(future)
+    def __init__(self, future: "asyncio.Future[T]") -> None:
```
Comment (pavelfeldman) [inline playwright/_impl/_async_base.py:42]: Let's only deal with cast here.

File: playwright/_impl/_async_base.py [hunk 8]
```diff
@@ -68,17 +73,19 @@ def _wrap_handler(self, handler: Any) -> Callable[..., None]:
             return mapping.wrap_handler(handler)
         return handler
 
-    def on(self, event: str, f: Any) -> None:
+    def on(self, event: str, f: Callable[..., Union[Awaitable[None], None]]) -> None:
         """Registers the function ``f`` to the event name ``event``."""
         self._impl_obj.on(event, self._wrap_handler(f))
 
-    def once(self, event: str, f: Any) -> None:
+    def once(self, event: str, f: Callable[..., Union[Awaitable[None], None]]) -> None:
```
Comment (pavelfeldman) [inline playwright/_impl/_async_base.py:80]: ditto

File: playwright/_impl/_async_base.py [hunk 9]
```diff
@@ -68,17 +73,19 @@ def _wrap_handler(self, handler: Any) -> Callable[..., None]:
...
         """The same as ``self.on``, except that the listener is automatically
         removed after being called.
         """
         self._impl_obj.once(event, self._wrap_handler(f))
 
-    def remove_listener(self, event: str, f: Any) -> None:
+    def remove_listener(
+        self, event: str, f: Callable[..., Union[Awaitable[None], None]]
```
Comment (pavelfeldman) [inline playwright/_impl/_async_base.py:87]: ditto

File: playwright/_impl/_connection.py (hunk 3 above)
Comment (pavelfeldman) [reply playwright/_impl/_connection.py:322]: You don't need this change to move to cast. It does not hurt, but this comment is misleading

File: playwright/_impl/_sync_base.py [hunk 10]
```diff
@@ -167,4 +172,7 @@ def __exit__(
         exc_val: BaseException,
         traceback: TracebackType,
     ) -> None:
-        self.close()  # type: ignore
+        self.close()
```
Comment (pavelfeldman) [inline playwright/_impl/_sync_base.py:175]: One thing at a time

File: playwright/_impl/_connection.py (hunk 3 above)
Comment (kumaraditya303) [reply playwright/_impl/_connection.py:322]: Updated comment

File: playwright/_impl/_async_base.py (hunk 5 above)
Comment (kumaraditya303) [reply playwright/_impl/_async_base.py:26]: Okay

File: playwright/_impl/_sync_base.py (hunk 10 above)
Comment (kumaraditya303) [reply playwright/_impl/_sync_base.py:175]: Okay

File: playwright/_impl/_artifact.py (hunk 4 above)
Comment (kumaraditya303) [reply playwright/_impl/_artifact.py:40]: Stream inherits from ChannelOwner and from_channel returns ChannelOwner hence it is happy

File: playwright/_impl/_async_base.py (hunk 1 above)
Comment (kumaraditya303) [reply playwright/_impl/_async_base.py:76]: Separated it to https://github.com/microsoft/playwright-python/pull/790

**Comment by pavelfeldman** [comment]
Please open it when it is ready for review, otherwise it is unclear whether you updated the patch or just replied to comments.

**Comment by kumaraditya303** [comment]
I has separated the async base changes to another PR and waiting here for it to get merged before rebasing. Here no change is required just a rebasing

**Pull request closed without merging**